### DISCUSSION TITLE
Which environment variables do we want to expose in configuration?

### DISCUSSION THREAD

**Discussion by cristiano-belloni** [jpmorganchase/modular#2231] [General] [answered]
Let's discuss what env variables we want to add to our Modular configuration in v4.

@AlbertoBrusa has identified an exhaustive list of all ENV variables we read in Modular as part of this PR - https://github.com/jpmorganchase/modular/pull/2227

We believe that only things that affect build/production should be configurable by file, while the rest should only be ENV_VARIABLES

| **Variable**                   | **Description**                                                                         | **CRA** | **Configurable** |
|--------------------------------|-----------------------------------------------------------------------------------------|---------|------------------|
| **NODE_ENV**                   |                                                                                         | No      |      No            |
| **NODE_PATH**                  |                                                                                         | No      |       No           |
| **HTTPS**                      | Development server in HTTPS mode                                                        | Yes     |      No      |
| **SSL_CRT_FILE** |         |   No    |   No    |  
| **SSL_KEY_FILE** |         |    No   |    No   |  
| **HOST**                       | Specify different host to localhost                                                     | Yes     | No                |
| **PORT**                       | Specify port to run on                                                                  | Yes     |   No            |
| **BROWSER**                    | Specify browser to override default                                                     | Yes     |  No                |
| **BROWSER_ARGS**               | Arguments passed to browser when opening                                                | Yes     |  No              |
| **SKIP_PREFLIGHT_CHECK**       | Skip all the modular repo checks                                                        | No      |        No          |
| **~~USE_MODULAR_WEBPACK~~**    | Removed                                                                                 | No      |        No          |
| **USE_MODULAR_ESBUILD**         | Use esbuild instead of Webpack                                                           | No      | Yes              |
| **EXTERNAL_CDN_TEMPLATE**      | URL template to query a CDN                                                             | No      | Yes              |
| **PUBLIC_URL**                 | Force assets to be referenced verbatim to the url you provide                           | Yes     |       Yes         |
| **EXTERNAL_BLOCK_LIST**        | Specifies which dependencies are allowed to be rewritten to the CDN. Default is none    | No      | Yes            |
| **EXTERNAL_ALLOW_LIST**        | Specifies which dependencies are not allowed to be rewritten to the CDN. Default is all | No      | Yes             |
| **MODULAR_LOGGER_DEBUG**       | Turn on Debug logs                                                                      | No      |        No          |
| **MODULAR_LOGGER_MUTE**        | Mute all logs                                                                           | No      |      No        |
| **BROWSERSLIST_ENV**            |     Pass env to browserlist command                                      | No      |       No        |
| **SKIP_MODULAR_STARTUP_CHECK** | Skip modular startup checks (Always skipped in CI)                            | No      |     No         |
| **NO_HEADLESS_TESTS**          | Force tests to **not** run headless                                                     | No      |       No           |
| **GIT_AUTHOR_EMAIL**           |                                                                                         | No      |       No         |
| **GIT_AUTHOR_NAME**            |                                                                                         | No      |       No         |
|**MODULAR_LINT_FIX**  |   | No  |    No  |
| **REACT_EDITOR** |         |    Yes   |  No     |  
| **VISUAL** |         |    No   |    No   |  
| **EDITOR** |         |  No     |   No    |  
| **WDS_SOCKET_PATH** |  When set, Create React App will run the development server with a custom websocket path for hot module reloading. Normally, webpack-dev-server defaults to /ws for the SockJS pathname. You may use this variable to start local development on more than one Create React App project at a time. See [webpack-dev-server documentation](https://webpack.js.org/configuration/dev-server/#devserversockpath) for more details.       |   Yes    |   No    |  
| **WDS_SOCKET_HOST** |    When set, Create React App will run the development server with a custom websocket hostname for hot module reloading. Normally, webpack-dev-server defaults to window.location.hostname for the SockJS hostname. You may use this variable to start local development on more than one Create React App project at a time. See [webpack-dev-server documentation](https://webpack.js.org/configuration/dev-server/#devserversockhost) for more details.     |    Yes   |   No    |  
| **WDS_SOCKET_PORT** |    When set, Create React App will run the development server with a custom websocket port for hot module reloading. Normally, webpack-dev-server defaults to window.location.port for the SockJS port. You may use this variable to start local development on more than one Create React App project at a time. See [webpack-dev-server documentation](https://webpack.js.org/configuration/dev-server/#devserversockport) for more details.     |    Yes   |   No    |  
| **DANGEROUSLY_DISABLE_HOST_CHECK** |         |  No     |    No   |  
| **INLINE_RUNTIME_CHUNK** |      Some apps do not need the benefits of saving a web request, so not inlining the chunk makes for a smoother build process   |   Yes    |    No   |  
| **GENERATE_SOURCEMAP** |     Source maps are resource heavy and can cause out of memory issue for large source files.  We need to make it configurable everywhere we currently set it    |    Yes   |    Yes   |  
| **IMAGE_INLINE_SIZE_LIMIT** |    This one is Webpack-only    |    Yes    |   No   |

**Comment by AlbertoBrusa** [comment]
To start off the discussion, I feel like anything that users might expect to be configurable coming from CRA should easily be configurable in modular through the config:
HTTPS
HOST
PORT
BROWSER
BROWSER_ARGS
PUBLIC_URL

**Comment by cristiano-belloni** [comment]
`EXTERNAL_BLOCK_LIST` and `EXTERNAL_ALLOW_LIST` should be configurable, with a default value of `undefined`

**Reply by AlbertoBrusa** [reply]
Agreed - I've kept the the default as  ['**'] for allow and [] for block as is already the case

**Reply by cristiano-belloni** [reply]
Even better!

**Comment by cristiano-belloni** [comment]
@AlbertoBrusa look for variables that are determined by destructuring `process.env` too; for example, `HTTPS` is useless without `SSL_CRT_FILE` and `SSL_KEY_FILE`:

```ts
const { SSL_CRT_FILE, SSL_KEY_FILE, HTTPS } = process.env;
```

**Reply by AlbertoBrusa** [reply]
Good catch, will add them in

**Reply by AlbertoBrusa** [reply]
Looking at the code HTTPS can be provided and is still useful without the other two, so we need to decide if we want to configure all three by the file, for now I've done HTTPS

**Comment by cristiano-belloni** [comment]
`SOCKER` -> `SOCKET`

**Comment by AlbertoBrusa** [comment, marked as answer]
Finalised list of configuration options we want to expose via a configuration file:

 | **Variable**                   | **Description**                                                                         | 
|--------------------------------|-----------------------------------------------------------------------------------------|
| **USE_MODULAR_ESBUILD**         | Use esbuild instead of Webpack                                                           | 
| **EXTERNAL_CDN_TEMPLATE**      | URL template to query a CDN                                                             | 
| **PUBLIC_URL**                 | Force assets to be referenced verbatim to the url you provide                           |
| **EXTERNAL_BLOCK_LIST**        | Specifies which dependencies are allowed to be rewritten to the CDN. Default is none. Discourage use    | 
| **EXTERNAL_ALLOW_LIST**        | Specifies which dependencies are not allowed to be rewritten to the CDN. Default is all. Discourage use| 
| **GENERATE_SOURCEMAP** |     Source maps are resource heavy and can cause out of memory issue for large source files.  We need to make it configurable everywhere we currently set it regardless of if it currently uses ENV variable    |
